buildpack.md

```diff
@@ -1123,6 +1123,8 @@ Each `key`:
 
 ## Deprecations
 This section describes all the features that are deprecated.
+Unless noted below, the lifecycle SHOULD fail if a buildpack declaring a newer Buildpack API tries to use features from an older Buildpack API.
+The lifecycle SHOULD warn and ignore if a buildpack declaring an older Buildpack API tries to use features from a newer Buildpack API.
```

Comment:
This looks like a lifecycle requirement "lifecycle SHOULD ...." and therefore probably doesn't belong in the buildpack API?


Do we actually want to specifcy this? I think we have been bitten a few times by unnecessarily over-specifying. If we need to warn buildpack authors we could say somethings more open ended like:
> The build MAY fail if the buildpack generates outputs that do not conform to the the declared API (e.g by setting undocumented keys in specified files). This is true even if the outputs would be valid under a different version of the buildpack API.

Reply:
Could we change it from "lifecycle SHOULD" to "lifecycle MAY" (so that it's directed to buildpack authors instead of lifecycle implementors)? 

I too am wary of over-specifying, but we have also just been bitten by under-specifying - because the spec didn't say anything about whether buildpacks could output both old and new sboms at the same time, people made assumptions and became unhappy later.

Reply:
On thinking about this further, maybe this information would be more helpful in the RFC template, to remind RFC authors when suggesting changes that this is how things will work by default unless they specify a migration path. It's a bit unfair to expect everyone to have seen this little footnote in the spec.

Reply:
My belief is that leaving things open-ended (_if mentioned at all_), thus leaving more flexibility for when details are clearer, is the correct move. People made assumptions because the corner case was never explicitly considered.

Although it's tempting to force all future corner cases to be considered ahead of time, because of a stipulation in spec or RFC, it's unreasonable to expect.
Although it's tempting to ward off future "unhappiness", by being able to point at a stipulation in spec or RFC, it's unreasonable to expect.

Reply:
Sounds good. I'm fine to close this if there's no appetite for it. Certainly having the spec be more open-ended on this topic is making our lives easier in the present moment.